aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp

```diff
@@ -1277,14 +1290,18 @@ struct ReductionMulOp {
 }  // namespace
 
 Tensor _sparse_csr_sum_cpu(const Tensor& input, IntArrayRef dims_to_sum, bool keepdim, c10::optional<ScalarType> dtype) {
-  ScalarType dtype_ = dtype.value_or(input.scalar_type());
+  ScalarType dtype_ = dtype.value_or(
+      at::isIntegralType(input.scalar_type(), /*includeBool=*/true)
+          ? ScalarType::Long
+          : input.scalar_type());
   Tensor input_ = input.to(dtype_);
```

Comment:
I'm surprised this is still needed. It seems like at least on the CPU the accumulation is happening in the required precision. Why upcast the input dtype when within the kernels you accumulate in the higher precision type before writing out the result?

Reply:
It is because, `sum.dim_IntList` of SparseCsr dispatch will call `_sparse_csr_sum.dim_dtype`, which output data type will follow input data type. Therefore, output dtype would be Long when input is integral.

Reply:
I think it is possible to eliminate `input.to(dtype_)` by incorporating the `isIntegralType` logic into `need_acc` logic in `reduce_sparse_csr_dim0_cpu_template`.
It will also require adjusting `reduce_sparse_csr_dim1_cpu_template` and `reduce_sparse_csr_dim01_cpu_template` so that these will return kLong tensor for integral types.

IIUC, the current state in this PR (as well as in the main branch) is less efficient (because `to` method call involves an unnecessary copy of input indices) than the desired state of not copying `input` tensor here rather than accessing the input tensor during the accumulation without the possible copy op of indices and values.

If @cpuhrsch agrees, these could be left as a follow-up task provided that the corresponding performance improvement issue is created.

Reply:
Yes, I agree @pearu.

I assumed this PR was also about performance, especially on CPU.

Aside from the performance benefits of folding the accumulate dtype into the calculation, it'll likely also make for simpler code.

Reply:
@yanbing-j could you please address this comment above.

Reply:
Done. I create `OpMathTypeWithIntegral` to include both for floating and integral types. So that low precision (bfloat16, float16 ..) will upcast to float, integral lower than int64_t will upcast to int64_t.

Could you please help review again? Thanks! @mingfeima @cpuhrsch @pearu